File: src/consts.ts
export interface Config {
    inputDir: string;
    owner: string;
    contentDir: string;
    pfpURL: string;
    baseURL: string;
    ignorePatterns: ['private']
    metadataImage: string
    pageTitle: string
    theme: {
        colors: {
            lightMode: Theme
            darkMode: Theme
        }
    }
}

interface Theme {
    background: string;
    lightbackground: string;
    darktext: string;
    heading: string;
    links: string;
    linkshover: string;
    text: string;
    comment: string;
}

export interface FileNode {
    name: string;
    path: string;
    type: "directory" | "file";
    children: FileNode[];
}

export interface SearchIndex {
    content: string;
    url: string;
}

export interface Metadata {
    title: string;
    date: string;
    tags: string[];
    author: string;
    category: string;
    status: "Draft" | "In Progress" | "Complete"
    priority: "Low" | "Medium" | "High"
    createdAt: string;
    updatedAt: string;
    description: string;
}

export interface tableOfContentsEntry {
    text: string,
    id: string,
    level: number
}
export const CONFIG_NAME = "grimoire.config.json"

File: src/consts.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

pub const GRIMOIRE_CONFIG_NAME: &'static str = "grimoire.config.json";

#[derive(Serialize, Debug, Clone, Deserialize)]
pub struct Config {
    pub project_name: String,
    pub content_dir: String,
    pub description: String,
    pub domain: String,
}

#[derive(Deserialize, Debug)]
pub struct FrontMatter {
    pub title: String,
    pub author: Option<String>,
    pub date: Option<String>,
    pub tags: Option<String>,
    pub description: Option<String>,
    pub slug: String,
    pub draft: Option<bool>,
    pub extra: Option<HashMap<String, serde_json::Value>>,
}
